### PR TITLE
Adding optimistic locking

### PR DESCRIPTION
**Pull request by MarvelGa**

### CODE REVIEW

**Review by jioter — DISMISSED** [review]

File: src/main/java/com/exadel/discountwebapp/discount/service/DiscountService.java [hunk 1]
```diff
@@ -43,9 +45,17 @@
     public DiscountResponseVO findById(Long id) {
         Discount discount = discountRepository.findById(id)
                 .orElseThrow(() -> new EntityNotFoundException(Discount.class, "id", id));
-        Long quantity = discount.getViewNumber() == null ? 1 : discount.getViewNumber() + 1;
-        discount.setViewNumber(quantity);
-        discountRepository.save(discount);
+        while (true) {
+            try {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/discount/service/DiscountService.java]: Before retrying it is preferable to read the entity again in a new transaction (with TransactionTemplate) one more time. Will discuss on our daily call

File: src/main/java/com/exadel/discountwebapp/discount/service/DiscountService.java [hunk 2]
```diff
@@ -43,9 +45,17 @@
...
+                Long quantity = discount.getViewNumber() == null ? 1 : discount.getViewNumber() + 1;
+                discount.setViewNumber(quantity);
+                discount.setVersion(discount.getVersion());
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/discount/service/DiscountService.java]: we shouldn't manage version field by ourselves, this field should be managed by Hibernate

File: src/main/resources/db/changelog/changes01/21072021_add_column_dis_version.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+                columnName: dis_version
+                tableName: discount
+                schemaName:  pulic
+      changes:
+        - addColumn:
+            tableName: discount
+            columns:
+              - column:
+                  name: dis_version
+                  type: BIGINT
```
Comment (ahlukhouski2) [inline src/main/resources/db/changelog/changes01/21072021_add_column_dis_version.yml:18]: let's set up the default version number as 1

File: src/main/java/com/exadel/discountwebapp/discount/service/DiscountService.java [hunk 4]
```diff
@@ -43,9 +45,17 @@
     public DiscountResponseVO findById(Long id) {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/discount/service/DiscountService.java]: let's rename this method to reflect the update of the Discount -> findByIdAndUpdateStatistics

File: src/main/resources/db/changelog/changes01/21072021_add_column_dis_version.yml (hunk 3 above)
Comment (MarvelGa) [reply src/main/resources/db/changelog/changes01/21072021_add_column_dis_version.yml:18]: fixed

File: src/main/java/com/exadel/discountwebapp/discount/service/DiscountService.java (hunk 4 above)
Comment (MarvelGa) [reply src/main/java/com/exadel/discountwebapp/discount/service/DiscountService.java]: fixed

**Review by ahlukhouski2 — APPROVED** [review]

**Review by Bozhokmaria — APPROVED** [review]

**Review by VolodymyrBushko — APPROVED** [review]

**Pull request merged**